### DISCUSSION TITLE
Using real_parent to remove the need for Path::canonicalize

### DISCUSSION THREAD

**Discussion by tesujimath** [nushell/nushell#12978] [General]
I asked the question in #12628 about Nushell's eagerness to canonicalize paths, and the ergonomic problems this causes.

Since then, I have seen some great work from @YizhePKU on removing canonicalize for PWD (#12603 and following). I love this, thank you!

An underlying reason to use canonicalized paths remains, however, and I decided to address that specifically.  The problem is, Rust's standard library `Path::parent` method is unusable on non-canonicalized paths, because it works purely lexically.  For example, `Path::new("..").parent() == ""` :exploding_head: 

I observe that the Rust community at large has embraced path canonicalization for this reason, and I think that has set us all on an unhelpful path.

So I wrote a [real_parent](https://docs.rs/real_parent/latest/real_parent/) crate, which does this correctly, without requiring canonicalized paths.

I am interested to use this to wean Nushell off canonicalized paths in general, for the reasons I give in my crate rationale.

The first step would be to transition from canonicalized paths to non-canonicalized but absolute ones, and to replace calls to `Path::parent` with `real_parent`.

Does this seem reasonable to others here?

**Comment by fdncred** [comment]
I'm not sure if using the `real_parent` crate is the right thing to do or not but after looking at your crate, I have concerns about it working properly in Windows. For instance, I don't see any tests with `DRIVE:\PATH` or `\\SERVER\SHARE` so it makes me think there will be Windows edge cases where it will not work as expected.

**Reply by tesujimath** [reply]
Hi Darren,

The way the tests work is that all test cases are run with both relative and absolute paths.  So when the tests are run on Windows (for which I have a GitHub action), this covers the DRIVE:\Path case.  See [here](https://github.com/tesujimath/real_parent/blob/main/tests/helpers/mod.rs#L205).

The point about the tests is that they can't make assumptions about any paths we haven't created as part of test setup.  So all paths will be within the link farm temp dir.  We can get their paths as absolute easily enough, and that is already done.

But I am unsure how to add automated tests for `\\SERVER\SHARE`, because I don't know how to generate such paths within the test suite. As far as I understand it, to make such a path work would require the user to have added a share on their computer.  If there's actually a way to do that I would be happy to add it.

**Reply by ChrisDenton** [reply]
`\\localhost\c$` is the UNC equivalent to `C:\`. The `localhost` share should be available by default.

**Reply by tesujimath** [reply]
Ah, thanks @ChrisDenton!  I'll try adding that to the tests for Windows.

**Reply by tesujimath** [reply]
I [extended the tests on Windows](https://github.com/tesujimath/real_parent/pull/6) to include UNC paths (thanks Chris for the pointer!)

I also made the tests output the paths they actually check, which you can see if you run the tests with `--nocapture`.  On Windows you get three paths per test case as follows:
```
verified "A/B/__c".real_parent() == "A\B\C"
verified "C:\Users\simon\AppData\Local\Temp\.tmpIro8Uy\A/B/__c".real_parent() == "C:\Users\simon\AppData\Local\Temp\.tmpIro8Uy\A\B\C"
verified "\\localhost\C$\Users\simon\AppData\Local\Temp\.tmpIro8Uy\A\B\__c".real_parent() == "\\localhost\C$\Users\simon\AppData\Local\Temp\.tmpIro8Uy\A\B\C"
```

It was a little tricky to cope with disk-path based absolute symlinks appearing in tests with UNC paths, but I made that work.

This may be enough to verify behaviour on Windows! :crossed_fingers:

**Reply by fdncred** [reply]
Did I miss a PR with real_parent? Last week some of the core-team looked at it and thought it was interesting and my work for us. I've been offline for a bit so I'm not sure if this has progressed.

**Reply by tesujimath** [reply]
Ah, no, sorry I haven't had a chance to progress this yet. I'm excited to try, and intend to run locally with such a change as soon as I can get to it!

I'll keep you posted.

**Reply by fdncred** [reply]
Thanks. No hurry. I just wanted to make sure we weren't the bottleneck.

**Reply by devyn** [reply]
This looks awesome!

**Comment by tesujimath** [comment]
Just to say, I haven't forgotten about this, have just been a bit busy.  I hope to have a PR next week.

Because I am having to change things which are conditionally compiled for `macos` and `windows` targets, I'm also taking a bit of time to get set up with cross-compilation and cross-platform rust-analyzer.  Seems worth it for the long haul!

**Reply by fdncred** [reply]
Next week is fine. We're coming upon a release again so that should be good timing right after a release to give plenty of testing time.

**Comment by tesujimath** [comment]
Hoping to follow up with the no-canonicalize tomorrow, but I realised I need to add `real_clean` to the real_parent crate, which is mostly done, just need a few more tests.
